python/ray/data/_internal/table_block.py

```diff
@@ -46,6 +48,20 @@ def add(self, item: Union[dict, TableRow, np.ndarray]) -> None:
                 "Returned elements of an TableBlock must be of type `dict`, "
                 "got {} (type {}).".format(item, type(item))
             )
+
+        item_columns_names = sorted(item.keys())
+        if self._sorted_columns_names:
+            # Check all added rows have same columns.
+            if item_columns_names != self._sorted_columns_names:
+                raise ValueError(
+                    "Current row has different columns compared to previous rows. "
+                    f"Columns of current row: {item_columns_names}, "
+                    f"Columns of previous rows: {self._sorted_columns_names}."
+                )
```

Comment:
This checks column names, but what if there are other arrow invalid issue such as incompatible value types? (string vs int)

Reply:
@ericl - good point, with:

```
def bad_fn(row):
  if random.random() > 0.5:
    return {"a": "hello1"}
  else:
    return {"a": 1}
    
ray.data.range(10).repartition(1).map(bad_fn)
```

It produces error message:

```
pyarrow.lib.ArrowInvalid: Could not convert 'hello1' with type str: tried to convert to int64
```

It seems already actionable to me with stack trace. From our side, we accumulate rows in a Python dictionary, then convert to an Arrow table. It would be hard for us to check each row's type when accumulating in Python dictionary here.

Full stack trace:
```
(_map_block_nosplit pid=80558) 2022-10-05 17:02:29,517	INFO worker.py:755 -- Task failed with retryable exception: TaskID(1e360ffa862f8fe3ffffffffffffffffffffffff01000000).
(_map_block_nosplit pid=80558) Traceback (most recent call last):
(_map_block_nosplit pid=80558)   File "python/ray/_raylet.pyx", line 781, in ray._raylet.execute_task
(_map_block_nosplit pid=80558)     with ray._private.worker._changeproctitle(title, next_title):
(_map_block_nosplit pid=80558)   File "python/ray/_raylet.pyx", line 785, in ray._raylet.execute_task
(_map_block_nosplit pid=80558)     outputs = function_executor(*args, **kwargs)
(_map_block_nosplit pid=80558)   File "/Users/chengsu/ray/python/ray/data/_internal/compute.py", line 439, in _map_block_nosplit
(_map_block_nosplit pid=80558)     for new_block in block_fn(block, *fn_args, **fn_kwargs):
(_map_block_nosplit pid=80558)   File "/Users/chengsu/ray/python/ray/data/dataset.py", line 303, in transform
(_map_block_nosplit pid=80558)     yield output_buffer.next()
(_map_block_nosplit pid=80558)   File "/Users/chengsu/ray/python/ray/data/_internal/output_buffer.py", line 74, in next
(_map_block_nosplit pid=80558)     block = self._buffer.build()
(_map_block_nosplit pid=80558)   File "/Users/chengsu/ray/python/ray/data/_internal/delegating_block_builder.py", line 64, in build
(_map_block_nosplit pid=80558)     return self._builder.build()
(_map_block_nosplit pid=80558)   File "/Users/chengsu/ray/python/ray/data/_internal/table_block.py", line 96, in build
(_map_block_nosplit pid=80558)     tables = [self._table_from_pydict(self._columns)]
(_map_block_nosplit pid=80558)   File "/Users/chengsu/ray/python/ray/data/_internal/arrow_block.py", line 111, in _table_from_pydict
(_map_block_nosplit pid=80558)     return pyarrow.Table.from_pydict(columns)
(_map_block_nosplit pid=80558)   File "pyarrow/table.pxi", line 1724, in pyarrow.lib.Table.from_pydict
(_map_block_nosplit pid=80558)   File "pyarrow/table.pxi", line 2368, in pyarrow.lib._from_pydict
(_map_block_nosplit pid=80558)   File "pyarrow/array.pxi", line 341, in pyarrow.lib.asarray
(_map_block_nosplit pid=80558)   File "pyarrow/array.pxi", line 315, in pyarrow.lib.array
(_map_block_nosplit pid=80558)   File "pyarrow/array.pxi", line 39, in pyarrow.lib._sequence_to_array
(_map_block_nosplit pid=80558)   File "pyarrow/error.pxi", line 143, in pyarrow.lib.pyarrow_internal_check_status
(_map_block_nosplit pid=80558)   File "pyarrow/error.pxi", line 99, in pyarrow.lib.check_status
(_map_block_nosplit pid=80558) pyarrow.lib.ArrowInvalid: Could not convert 'hello1' with type str: tried to convert to int64
```